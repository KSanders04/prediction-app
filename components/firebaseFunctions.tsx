import { auth, db, storage } from "../firebaseConfig";
import { 
  signInWithEmailAndPassword, 
  createUserWithEmailAndPassword, 
  getAuth, 
  reauthenticateWithCredential, 
  EmailAuthProvider, 
  updatePassword,
  } from "firebase/auth";
import { ref, uploadBytes, getDownloadURL } from 'firebase/storage';
import { doc, setDoc, getDoc, query, collection, where, getDocs, updateDoc } from "firebase/firestore";

{/*---- LOGIN FIREBASE/LOGIC ----*/}
// Handles user sign in with email and password
export const emailSignIn = async (email: string, password: string) => {
  try {
    const user = await signInWithEmailAndPassword(auth, email, password);
    const userCredential = user.user;
    const userRef = doc(db, "users", userCredential.uid);
    const userSnap = await getDoc(userRef);

    // If user doc doesn't exist, create it with default values
    if (!userSnap.exists()) {
      await setDoc(userRef, {
        uid: userCredential.uid,
        email: userCredential.email,
        createdAt: new Date().toISOString(),
        isGamemaster: null,
        correctPredictions: 0,
        gamesPlayed: 0,
        totalPoints: 0,
        totalPredictions: 0,
      });
    }
    return user;
  } catch (error: any) {
    throw error;
  }
};

{/*---- CREATE ACCOUNT FIREBASE/LOGIC ----*/}
// Handles user sign up with email, password, and profile info
export const emailSignUp = async (
  email: string,
  password: string,
  firstName: string,
  lastName: string,
  userName: string
) => {
  // Username validation
  if (!userName.trim()) throw new Error("Username cannot be empty.");
  if (userName.trim().length < 6) throw new Error("Username must be at least 6 characters long.");
  if (!firstName.trim() || !lastName.trim()) throw new Error("First and last name cannot be empty.");
  if (!email.trim() || !password.trim()) throw new Error("Email and password cannot be empty.");
  if (password.length < 6) throw new Error("Password must be at least 6 characters long.");

  // Check if username exists
  const usernameQuery = query(collection(db, "users"), where("userName", "==", userName));
  const usernameSnapshot = await getDocs(usernameQuery);
  if (!usernameSnapshot.empty) throw new Error("That username is already taken. Please choose another.");

  // Check if email exists
  const emailQuery = query(collection(db, "users"), where("email", "==", email));
  const emailSnapshot = await getDocs(emailQuery);
  if (!emailSnapshot.empty) throw new Error("That email is already in use. Please choose another.");

  // Create account
  const user = await createUserWithEmailAndPassword(auth, email, password);
  const userCredential = user.user;
  const userRef = doc(db, "users", userCredential.uid);
  const userSnap = await getDoc(userRef);

  if (!userSnap.exists()) {
    await setDoc(userRef, {
      uid: userCredential.uid,
      firstName,
      lastName,
      userName,
      email: userCredential.email,
      createdAt: new Date().toISOString(),
      isAdmin: null,
      correctPredictions: 0,
      gamesPlayed: 0,
      totalPoints: 0,
      totalPredictions: 0,
    });
  }

  return user;
};

{/*---- SIGN OUT FIREBASE/LOGIC ----*/}
// Handles user sign out
export const signOutUser = async () => {
  await auth.signOut();
};
// Listen for auth state changes and redirect if not logged in
export const listenForSignOut = (redirect: () => void) => {
  getAuth().onAuthStateChanged((user) => {
    if (!user) redirect();
  });
};

{/*---- PROFILE FIREBASE/LOGIC ----*/}
// Get current user
export const getCurrentUser = () => {
  return auth.currentUser;
};

// Fetch user data
export const fetchUserData = async (uid: string) => {
  const docRef = doc(db, 'users', uid);
  const docSnap = await getDoc(docRef);
  return docSnap.exists() ? docSnap.data() : null;
};

// Upload profile image and update Firestore
export const uploadProfileImage = async (uid: string, uri: string) => {
  const response = await fetch(uri);
  const blob = await response.blob();
  const storageRef = ref(storage, `profilePics/${uid}.jpg`);
  await uploadBytes(storageRef, blob);
  const downloadURL = await getDownloadURL(storageRef);
  await updateDoc(doc(db, 'users', uid), { profilePic: downloadURL });
  return downloadURL;
};

// Update username
export const updateUsername = async (uid: string, newUsername: string) => {
  await updateDoc(doc(db, 'users', uid), { userName: newUsername });
};

// Get current user rank (stub, replace with real logic if needed)
export const getCurrentUserRank = () => 1;

// Get rank suffix
export const getRankSuffix = (rank: number) => {
  if (rank % 100 >= 11 && rank % 100 <= 13) return 'th';
  switch (rank % 10) {
    case 1: return 'st';
    case 2: return 'nd';
    case 3: return 'rd';
    default: return 'th';
  }
};

// Calculate accuracy percentage
export const getAccuracy = (correct: number, total: number) =>
  total > 0 ? Math.round((correct / total) * 100) : 0;

// Format last played date
export const formatLastPlayed = (timestamp: any) => {
  if (!timestamp) return 'Never';
  const date = timestamp.toDate ? timestamp.toDate() : new Date(timestamp);
  return date.toLocaleDateString();
};

{/*---- CHANGE PASSWORD FIREBASE/LOGIC ----*/}
// Change password logic for use in changePasswordPage
export const changePassword = async (
  currentPassword: string,
  newPassword: string
) => {
  const user = auth.currentUser;
  if (!user || !user.email) {
    throw new Error('No authenticated user.');
  }
  const credential = EmailAuthProvider.credential(user.email, currentPassword);
  await reauthenticateWithCredential(user, credential);
  await updatePassword(user, newPassword);
};