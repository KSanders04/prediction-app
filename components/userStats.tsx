import { View, Text, StyleSheet } from "react-native";
import React from "react";



interface User {
  userName: string,
  totalPoints: number;
  correctPredictions: number;
  totalPredictions: number;
  gamesPlayed: number;
  lastPlayed: any;
}

export const UserStats: React.FC<{
  currentUser: User;
  authUser: any;
  getCurrentUserRank: () => number;
  getRankSuffix: (rank: number) => string;
  getAccuracy: (correct: number, total: number) => number;
  formatLastPlayed: (timestamp: any) => string;
}> = ({
  currentUser,
  authUser,
  getCurrentUserRank,
  getRankSuffix,
  getAccuracy,
  formatLastPlayed,
}) => (
  <View style={styles.currentUserSection}>
   <Text style={styles.sectionTitle}>üìä Your Stats</Text>

<View style={styles.nameContainer}>
  <Text style={styles.nameText}>
    {currentUser.userName || authUser?.email || `User_${authUser?.uid?.slice(0, 6)}`}
  </Text>
  {currentUser.userName && (
  <Text style={styles.usernameText}>@{currentUser.userName}</Text>
)}

</View>
    <View style={styles.statsGrid}>
      <View style={styles.statCard}>
        <Text style={styles.statValue}>{getCurrentUserRank() || '--'}</Text>
        <Text style={styles.statLabel}>
          {getCurrentUserRank()
            ? `${getCurrentUserRank()}${getRankSuffix(getCurrentUserRank())} Place`
            : 'Unranked'}
        </Text>
      </View>
      <View style={styles.statCard}>
        <Text style={styles.statValue}>{currentUser.totalPoints}</Text>
        <Text style={styles.statLabel}>Total Points</Text>
      </View>
      <View style={styles.statCard}>
        <Text style={styles.statValue}>
          {getAccuracy(currentUser.correctPredictions, currentUser.totalPredictions)}%
        </Text>
        <Text style={styles.statLabel}>Accuracy</Text>
      </View>
      <View style={styles.statCard}>
        <Text style={styles.statValue}>{currentUser.gamesPlayed}</Text>
        <Text style={styles.statLabel}>Games Played</Text>
      </View>
    </View>
    <View style={styles.detailedStats}>
      <Text style={styles.detailText}>
        ‚úÖ Correct Predictions: {currentUser.correctPredictions} / {currentUser.totalPredictions}
      </Text>
      <Text style={styles.detailText}>
        üïê Last Played: {formatLastPlayed(currentUser.lastPlayed)}
      </Text>
    </View>
  </View>
);

const styles = StyleSheet.create({
  currentUserSection: {
    backgroundColor: 'white',
    margin: 20,
    padding: 20,
    borderRadius: 15,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 8,
    elevation: 5,
    borderLeftWidth: 4,
    borderLeftColor: '#3498db',
  },
  sectionTitle: {
    fontSize: 20,
    fontWeight: 'bold',
    marginBottom: 15,
    color: '#2c3e50',
  },
  userEmail: {
    fontSize: 14,
    color: '#7f8c8d',
    textAlign: 'center',
    marginBottom: 15,
    fontStyle: 'italic',
  },
  statsGrid: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'space-between',
    marginBottom: 15,
  },
  statCard: {
    width: '48%',
    backgroundColor: '#f8f9fa',
    padding: 15,
    borderRadius: 12,
    alignItems: 'center',
    marginBottom: 10,
  },
  statValue: {
    fontSize: 24,
    fontWeight: 'bold',
    color: '#2c3e50',
  },
  statLabel: {
    fontSize: 12,
    color: '#7f8c8d',
    marginTop: 4,
    textAlign: 'center',
  },
  detailedStats: {
    borderTopWidth: 1,
    borderTopColor: '#ecf0f1',
    paddingTop: 15,
  },
  detailText: {
    fontSize: 14,
    color: '#34495e',
    marginBottom: 5,
  },
  nameContainer: {
  alignItems: 'center',
  marginBottom: 15,
},
nameText: {
  fontSize: 18,
  fontWeight: '600',
  color: '#2c3e50',
},
usernameText: {
  fontSize: 14,
  color: '#7f8c8d',
  fontStyle: 'italic',
  marginTop: 2,
},
});