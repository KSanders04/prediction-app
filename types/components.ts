import { User } from './user';

export interface UserStatsProps {
  currentUser: User;
  authUser: any;
  getCurrentUserRank: () => number;
  getRankSuffix: (rank: number) => string;
  getAccuracy: (correct: number, total: number) => number;
  formatLastPlayed: (timestamp: any) => string;
}

export interface ProfileStatsProps {
  currentUser: User;
  authUser: any;
  getCurrentUserRank: () => number;
  getRankSuffix: (rank: number) => string;
  getAccuracy: (correct: number, total: number) => number;
  formatLastPlayed: (timestamp: any) => string;
}

export interface ActivePlayersProps {
  currentGameId?: string | null;
}