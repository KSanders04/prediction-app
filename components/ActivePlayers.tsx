// components/ActivePlayers.tsx
import React, { useState, useEffect } from 'react';
import { View, Text, StyleSheet, ScrollView } from 'react-native';
import { Timestamp } from 'firebase/firestore';
import { listenForActivePlayers, formatTimeAgo } from './firebaseFunctions';


import { ActivePlayer, ActivePlayersProps } from '@/types';

export const ActivePlayers: React.FC<ActivePlayersProps> = ({ currentGameId }) => {
  const [activePlayers, setActivePlayers] = useState<ActivePlayer[]>([]);
  const [totalViewers, setTotalViewers] = useState<number>(0);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    if (!currentGameId) {
      setActivePlayers([]);
      setTotalViewers(0);
      setLoading(false);
      return;
    }

    setLoading(true);
    const unsubscribe = listenForActivePlayers(
      currentGameId,
      (players, viewers) => {
        setActivePlayers(players);
        setTotalViewers(viewers);
        setLoading(false);
      },
      (error) => {
        console.error('Error listening to active players:', error);
        setLoading(false);
      }
    );

    return () => { if (unsubscribe) unsubscribe(); };
  }, [currentGameId]);

  if (loading) {
    return (
      <View style={styles.container}>
        <Text style={styles.title}>ðŸ‘¥ Live Activity</Text>
        <Text style={styles.loadingText}>Loading active players...</Text>
      </View>
    );
  }

  if (!currentGameId) {
    return (
      <View style={styles.container}>
        <Text style={styles.title}>ðŸ‘¥ Live Activity</Text>
        <Text style={styles.emptyText}>No active game. Join a game to see live activity!</Text>
      </View>
    );
  }

  const playingCount = activePlayers.filter(player => player.isPlaying).length;
  const watchingCount = totalViewers - playingCount;

  return (
    <View style={styles.container}>
      <Text style={styles.title}>ðŸ‘¥ Live Activity</Text>

      {/* Summary Stats */}
      <View style={styles.summaryContainer}>
        <View style={styles.summaryCard}>
          <Text style={styles.summaryNumber}>{totalViewers}</Text>
          <Text style={styles.summaryLabel}>Total Viewers</Text>
        </View>
        <View style={styles.summaryCard}>
          <Text style={[styles.summaryNumber, styles.playingNumber]}>{playingCount}</Text>
          <Text style={styles.summaryLabel}>Active Players</Text>
        </View>
        <View style={styles.summaryCard}>
          <Text style={styles.summaryNumber}>{watchingCount}</Text>
          <Text style={styles.summaryLabel}>Just Watching</Text>
        </View>
      </View>

      {/* Active Players List */}
      {activePlayers.length === 0 ? (
        <Text style={styles.emptyText}>No one is currently active in this game.</Text>
      ) : (
        <ScrollView style={styles.playersList} showsVerticalScrollIndicator={false}>
          <Text style={styles.sectionTitle}>ðŸŸ¢ Currently Online</Text>
          {activePlayers.map((player, index) => (
            <View key={`${player.userId}-${index}`} style={[
              styles.playerItem,
              player.isPlaying && styles.playingPlayerItem
            ]}>
              <View style={styles.playerInfo}>
                <Text style={styles.playerEmail}>
                  {player.isPlaying ? 'ðŸŽ¯' : 'ðŸ‘€'} {player.email || 'Anonymous'}
                </Text>
                <Text style={styles.playerStatus}>
                  {player.isPlaying ? 'Playing' : 'Watching'} â€¢ Joined {formatTimeAgo(player.joinedAt)}
                </Text>
              </View>
              <View style={[
                styles.statusDot,
                player.isPlaying ? styles.playingDot : styles.watchingDot
              ]} />
            </View>
          ))}
        </ScrollView>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'white',
    margin: 20,
    padding: 20,
    borderRadius: 15,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 8,
    elevation: 5,
    maxHeight: 400,
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
    marginBottom: 15,
    color: '#2c3e50',
    textAlign: 'center',
  },
  loadingText: {
    textAlign: 'center',
    color: '#7f8c8d',
    fontStyle: 'italic',
    padding: 20,
  },
  emptyText: {
    textAlign: 'center',
    color: '#7f8c8d',
    fontStyle: 'italic',
    padding: 20,
    fontSize: 14,
  },
  summaryContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: 20,
  },
  summaryCard: {
    flex: 1,
    backgroundColor: '#f8f9fa',
    padding: 12,
    borderRadius: 10,
    alignItems: 'center',
    marginHorizontal: 4,
  },
  summaryNumber: {
    fontSize: 20,
    fontWeight: 'bold',
    color: '#2c3e50',
  },
  playingNumber: {
    color: '#27ae60',
  },
  summaryLabel: {
    fontSize: 10,
    color: '#7f8c8d',
    marginTop: 2,
    textAlign: 'center',
  },
  sectionTitle: {
    fontSize: 16,
    fontWeight: '600',
    color: '#2c3e50',
    marginBottom: 10,
  },
  playersList: {
    maxHeight: 200,
  },
  playerItem: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 8,
    paddingHorizontal: 12,
    marginBottom: 6,
    backgroundColor: '#f8f9fa',
    borderRadius: 8,
    borderLeftWidth: 3,
    borderLeftColor: '#3498db',
  },
  playingPlayerItem: {
    backgroundColor: '#e8f5e8',
    borderLeftColor: '#27ae60',
  },
  playerInfo: {
    flex: 1,
  },
  playerEmail: {
    fontSize: 14,
    fontWeight: '500',
    color: '#2c3e50',
    marginBottom: 2,
  },
  playerStatus: {
    fontSize: 11,
    color: '#7f8c8d',
  },
  statusDot: {
    width: 8,
    height: 8,
    borderRadius: 4,
    marginLeft: 8,
  },
  playingDot: {
    backgroundColor: '#27ae60',
  },
  watchingDot: {
    backgroundColor: '#3498db',
  },
});