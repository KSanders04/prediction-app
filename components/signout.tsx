import { StyleSheet, TouchableOpacity, Text, Alert, View } from 'react-native';
import { router } from 'expo-router';
import { signOutUser, listenForSignOut } from '@/components/firebaseFunctions';
import FontAwesome from '@expo/vector-icons/FontAwesome';

export default function signout() {
  // Redirect to login if not authenticated
  listenForSignOut(() => router.replace('/'));

  const handleSignOut = async () => {
    Alert.alert(
      "Sign Out",
      "Are you sure you want to sign out?",
      [
        { text: "Cancel", style: "cancel" },
        { text: "Sign Out", style: "destructive", onPress: async () => {
            await signOutUser();
            router.replace('/');
          }
        }
      ]
    );
  };

  return (
    <TouchableOpacity style={styles.button} onPress={handleSignOut}>
    <View style={{ flexDirection: 'row', alignItems: 'center' }}>
      <FontAwesome
        name="sign-out"
        size={25}
        color="#FFFFFF"
        style={{ marginRight: 8, opacity: 1 }}
      />
      <Text style={styles.text}>Sign Out</Text>
    </View>
  </TouchableOpacity>
);
}

const styles = StyleSheet.create({
  button: {
    width: '90%',
    backgroundColor: '#3498db', 
    padding: 20,
    borderRadius: 15, 
    alignItems: 'center',
    justifyContent: 'center',
    shadowColor: '#5C6BC0', 
    shadowOffset: { width: 0, height: 4 },
    shadowOpacity: 0.4,
    shadowRadius: 5,
    elevation: 5, 
    marginTop: 10, 
    marginBottom: 20,
  },
  text: {
    color: '#FFFFFF', 
    fontSize: 18, 
    fontWeight: '600', 
  }
});