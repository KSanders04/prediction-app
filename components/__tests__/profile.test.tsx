import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import Profile from '@/app/(tabs)/profile';
import * as firebaseFunctions from '../../components/firebaseFunctions';

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({ navigate: jest.fn() }),
}));

jest.mock('prediction-app/components/signout.tsx', () => () => <></>);

jest.mock('../../components/firebaseFunctions', () => ({
  fetchUserData: jest.fn(),
  getCurrentUser: jest.fn(),
  updateUsername: jest.fn(),
  changePassword: jest.fn(),
  uploadProfileImage: jest.fn(),
  getCurrentUserRank: jest.fn(),
  getRankSuffix: jest.fn(),
  getAccuracy: jest.fn(),
  formatLastPlayed: jest.fn(),
}));

describe('Profile Screen', () => {
  const mockUser = { uid: '123456', email: 'test@example.com' };

  const mockData = {
    firstName: 'Jane',
    lastName: 'Doe',
    userName: 'jdoe',
    profilePic: '',
    totalPoints: 100,
    correctPrictions: 100,
    gamesPlayedictions: 80,
    totalPreded: 10,
    lastPlayed: null,
    groups: [],
  };

  beforeEach(() => {
    jest.clearAllMocks();
    (firebaseFunctions.getCurrentUser as jest.Mock).mockResolvedValue(mockUser);
    (firebaseFunctions.fetchUserData as jest.Mock).mockResolvedValue(mockData);
  });

  it('renders profile with user data', async () => {
    const { getByText } = render(<Profile />);
    await waitFor(() => {
      expect(getByText('Jane Doe')).toBeTruthy();
      expect(getByText('Edit Username')).toBeTruthy();
    });
  });

  it('lets user edit and save username', async () => {
    const { getByText, getByPlaceholderText } = render(<Profile />);
    await waitFor(() => getByText('Edit Username'));

    fireEvent.press(getByText('Edit Username'));

    const input = getByPlaceholderText('Enter new username');
    fireEvent.changeText(input, 'newUsername');
    fireEvent.press(getByText('Save'));

    await waitFor(() =>
      expect(firebaseFunctions.updateUsername).toHaveBeenCalledWith(
        mockUser.uid,
        'newUsername'
      )
    );
  });

  it('renders Change Password button and responds to press', async () => {
      const { getByText } = render(<Profile />);

      const button = await waitFor(() => getByText('Change Password'));
      expect(button).toBeTruthy();

      //This test shows that button does press and that no crashes happen
      fireEvent.press(button);
  });

    it('matches snapshot', () => {
      const tree = render(<Profile />).toJSON();
      expect(tree).toMatchSnapshot();
  });
});