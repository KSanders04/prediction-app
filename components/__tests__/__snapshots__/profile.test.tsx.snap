// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Screen matches snapshot 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "alignItems": "center",
      "justifyContent": "flex-start",
    }
  }
  style={
    {
      "backgroundColor": "#f5f7fa",
      "flex": 1,
      "padding": 20,
      "paddingTop": 60,
    }
  }
>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          {
            "testUri": "../../../assets/images/head_alizarin.png",
          }
        }
        style={
          {
            "backgroundColor": "#ecf0f1",
            "borderRadius": 60,
            "height": 120,
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": -55,
            "width": 120,
          }
        }
      />
    </View>
    <View
      style={
        {
          "alignItems": "flex-start",
          "marginBottom": -10,
        }
      }
    >
      <Text
        style={
          {
            "color": "#2c3e50",
            "fontSize": 24,
            "fontWeight": "bold",
            "paddingTop": 10,
          }
        }
      >
         
      </Text>
    </View>
    <View
      style={
        [
          {
            "flexDirection": "row",
            "justifyContent": "center",
          },
          {
            "marginBottom": -10,
            "marginTop": 35,
          },
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignSelf": "center",
            "backgroundColor": "#3498db",
            "borderRadius": 5,
            "height": 32,
            "justifyContent": "center",
            "marginLeft": 10,
            "opacity": 1,
            "paddingHorizontal": 12,
            "paddingVertical": 6,
          }
        }
      >
        <Text
          style={
            {
              "color": "#fff",
              "fontSize": 14,
              "fontWeight": "600",
            }
          }
        >
          Edit Username
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignSelf": "center",
            "backgroundColor": "#3498db",
            "borderRadius": 5,
            "height": 32,
            "justifyContent": "center",
            "marginLeft": 10,
            "opacity": 1,
            "paddingHorizontal": 12,
            "paddingVertical": 6,
          }
        }
      >
        <Text
          style={
            {
              "color": "#fff",
              "fontSize": 14,
              "fontWeight": "600",
            }
          }
        >
          Change Password
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "center",
          "marginBottom": -10,
          "marginTop": 20,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "marginHorizontal": 15,
          }
        }
      >
        <Text
          style={
            {
              "color": "#2c3e50",
              "fontSize": 18,
              "fontWeight": "bold",
              "marginTop": 10,
            }
          }
        >
          0
        </Text>
        <Text
          style={
            {
              "color": "black",
              "fontSize": 14,
            }
          }
        >
          Friends
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "marginHorizontal": 15,
          }
        }
      >
        <Text
          style={
            {
              "color": "#2c3e50",
              "fontSize": 18,
              "fontWeight": "bold",
              "marginTop": 10,
            }
          }
        >
          0
        </Text>
        <Text
          style={
            {
              "color": "black",
              "fontSize": 14,
            }
          }
        >
          Groups
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "marginTop": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "backgroundColor": "white",
            "borderLeftColor": "#3498db",
            "borderLeftWidth": 4,
            "borderRadius": 15,
            "elevation": 5,
            "margin": 20,
            "padding": 20,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.1,
            "shadowRadius": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#2c3e50",
              "fontSize": 20,
              "fontWeight": "bold",
              "marginBottom": 15,
            }
          }
        >
          📊 Your Stats
        </Text>
        <View
          style={
            {
              "alignItems": "center",
              "marginBottom": 15,
            }
          }
        >
          <Text
            style={
              {
                "color": "#2c3e50",
                "fontSize": 18,
                "fontWeight": "600",
              }
            }
          >
            User_undefined
          </Text>
        </View>
        <View
          style={
            {
              "flexDirection": "row",
              "flexWrap": "wrap",
              "justifyContent": "space-between",
              "marginBottom": 15,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#f8f9fa",
                "borderRadius": 12,
                "marginBottom": 10,
                "padding": 15,
                "width": "48%",
              }
            }
          >
            <Text
              style={
                {
                  "color": "#2c3e50",
                  "fontSize": 24,
                  "fontWeight": "bold",
                }
              }
            >
              %
            </Text>
            <Text
              style={
                {
                  "color": "#7f8c8d",
                  "fontSize": 12,
                  "marginTop": 4,
                  "textAlign": "center",
                }
              }
            >
              Accuracy
            </Text>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#f8f9fa",
                "borderRadius": 12,
                "marginBottom": 10,
                "padding": 15,
                "width": "48%",
              }
            }
          >
            <Text
              style={
                {
                  "color": "#2c3e50",
                  "fontSize": 24,
                  "fontWeight": "bold",
                }
              }
            >
              0
            </Text>
            <Text
              style={
                {
                  "color": "#7f8c8d",
                  "fontSize": 12,
                  "marginTop": 4,
                  "textAlign": "center",
                }
              }
            >
              Games Played
            </Text>
          </View>
        </View>
        <View
          style={
            {
              "borderTopColor": "#ecf0f1",
              "borderTopWidth": 1,
              "paddingTop": 15,
            }
          }
        >
          <Text
            style={
              {
                "color": "#34495e",
                "fontSize": 14,
                "marginBottom": 5,
              }
            }
          >
            ✅ Correct Predictions: 
            0
             / 
            0
          </Text>
          <Text
            style={
              {
                "color": "#34495e",
                "fontSize": 14,
                "marginBottom": 5,
              }
            }
          >
            🕐 Last Played: 
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
